mm2src/mm2_bin_lib/src/mm2_native_lib.rs

```diff
@@ -48,40 +40,102 @@ pub unsafe extern "C" fn mm2_main(conf: *const c_char, log_cb: extern "C" fn(lin
         }};
     }
 
-    if LP_MAIN_RUNNING.load(Ordering::Relaxed) {
-        eret!(MainErr::AlreadyRuns)
-    }
-    CTX.store(0, Ordering::Relaxed); // Remove the old context ID during restarts.
-
     if conf.is_null() {
-        eret!(MainErr::ConfIsNull)
+        eret!(StartupResultCode::InvalidParams, "Configuration is null")
     }
-    let conf = CStr::from_ptr(conf);
-    let conf = match conf.to_str() {
+    let conf_cstr = CStr::from_ptr(conf);
+    let conf_str = match conf_cstr.to_str() {
         Ok(s) => s,
-        Err(e) => eret!(MainErr::ConfNotUtf8, e),
+        Err(e) => eret!(
+            StartupResultCode::InvalidParams,
+            format!("Configuration is not valid UTF-8: {}", e)
+        ),
+    };
+
+    let conf: json::Value = match json::from_str(conf_str) {
+        Ok(v) => v,
+        Err(e) => eret!(
+            StartupResultCode::ConfigError,
+            format!("Failed to parse configuration: {}", e)
+        ),
     };
-    let conf = conf.to_owned();
+
+    if let Err(true) = LP_MAIN_RUNNING.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed) {
+        eret!(StartupResultCode::AlreadyRunning, "MM2 is already running");
+    }
+
+    // Remove the old context ID during restarts.
+    CTX.store(0, Ordering::Relaxed);
 
     register_callback(FfiCallback::with_ffi_function(log_cb));
+
+    let (init_sender, init_receiver) = std::sync::mpsc::channel();
+
     let rc = thread::Builder::new().name("lp_main".into()).spawn(move || {
-        if let Err(true) = LP_MAIN_RUNNING.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed) {
-            log!("lp_main already started!");
-            return;
-        }
         let ctx_cb = &|ctx| CTX.store(ctx, Ordering::Relaxed);
-        match catch_unwind(move || mm2_main::run_lp_main(Some(&conf), ctx_cb, KDF_VERSION.into(), KDF_DATETIME.into()))
-        {
-            Ok(Ok(_)) => log!("run_lp_main finished"),
-            Ok(Err(err)) => log!("run_lp_main error: {}", err),
-            Err(err) => log!("run_lp_main panic: {:?}", any_to_str(&*err)),
+
+        let inner_sender = init_sender.clone();
+
+        match catch_unwind(move || {
+            let params = LpMainParams::with_conf(conf).log_filter(None);
+
+            match block_on(mm2_main::lp_main(
```

Comment:
stray thought: why do we run lp_main in another thread?
since it's short lived we can run it outside and not use that channel thing. no?

Reply:
Done, please give it a look